include/util/json_renderer.hpp

```diff
@@ -4,191 +4,164 @@
 #ifndef JSON_RENDERER_HPP
 #define JSON_RENDERER_HPP
 
-#include "util/cast.hpp"
-#include "util/ieee754.hpp"
 #include "util/string_util.hpp"
 
 #include "osrm/json_container.hpp"
 
+#include <algorithm>
 #include <iterator>
 #include <ostream>
 #include <string>
 #include <vector>
 
+#include <fmt/compile.h>
+
 namespace osrm
 {
 namespace util
 {
 namespace json
 {
 
-namespace
-{
-constexpr int MAX_FLOAT_STRING_LENGTH = 256;
-}
-
-struct Renderer
+template <typename Out> struct Renderer
 {
-    explicit Renderer(std::ostream &_out) : out(_out) {}
+    explicit Renderer(Out &_out) : out(_out) {}
 
-    void operator()(const String &string) const
+    void operator()(const String &string)
     {
-        out << "\"";
-        out << escape_JSON(string.value);
-        out << "\"";
+        write('"');
+        // here we assume that vast majority of strings don't need to be escaped,
+        // so we check it first and escape only if needed
+        auto size = SizeOfEscapedJSONString(string.value);
```

Comment:
Can we instead just search for the existence of a character that requires escaping?

Reply:
Do we have a feel for how often OSRM responses do require escaping? I.e how representative are these benchmarks?

Reply:
> Can we instead just search for the existence of a character that requires escaping?

We can, but current implementation allows to preallocate string for escaped string if we need it(because we know required size). ANd also assuming that vast majority of string don’t need to be escaped(I.e. don’t contain special characters) we will have to scan the whole string most of the times anyway.

> Do we have a feel for how often OSRM responses do require escaping?

That’s the great question. I don’t have numbers, just a gut feeling… Will try to collect some statistics from real responses.

Btw this idea is inspired by implementation in https://github.com/nlohmann/json

Reply:
Actually probably should be a good idea to take huge real JSON response, convert it to our json:: objects using some script and try to check some of optimizations on it.

Reply:
Well, actually I found 2 "sources" of escaped strings in responses.

1. Geometries(tbh I didn't know we can have slashes there and they appear quite often...)
<img width="1007" alt="Screenshot 2022-10-02 at 22 37 42" src="https://user-images.githubusercontent.com/266271/193475380-9b88a3ba-f670-461f-8595-51298088fb6f.png">

But even though they appear quite often these strings are quite big usually and it should allow us to preallocate string buffer for them(but we have to scan string twice)...

2. Road names and pronunciations. I quickly hacked osrm-extract to count number of `name` and `name:pronunciation` tags which need to be escaped in osm.pbf and got the following numbers for Croatia and Poland respectively:
<img width="1127" alt="Screenshot 2022-10-02 at 22 10 58" src="https://user-images.githubusercontent.com/266271/193475457-8ca2d969-3ec6-481b-a3ea-d0644009c4b7.png">
<img width="472" alt="Screenshot 2022-10-02 at 22 37 02" src="https://user-images.githubusercontent.com/266271/193475458-45926a26-f129-410d-a0a7-6fb79b1e795a.png">

After that I updated benchmark to use real output of OSRM(huge route from Portugal to Korea) and "played" a bit with this optimisation. And my conclusion is that it seems this optimisation is still useful on real JSONs, but improvement is barely noticeable.

### Without this optimisation (i.e. just always call `EscapeJSONString`)
<img width="1166" alt="Screenshot 2022-10-02 at 22 38 44" src="https://user-images.githubusercontent.com/266271/193475560-507bb342-00fb-42d0-be9a-d0858fdfd9b2.png">

### With optimisation
<img width="852" alt="Screenshot 2022-10-02 at 22 38 23" src="https://user-images.githubusercontent.com/266271/193475567-396c7d8e-89e4-4356-b0b5-ace3a6a47083.png">

So I don't have strong opinion here: we can remove it(less code == more readable code) or leave it as is since it seems to make things slightly better.

WDYT?

Reply:
> Can we instead just search for the existence of a character that requires escaping?

Decided to check this variant(i.e. without preallocation if string requires escaping) and it surprisingly even more better on this particular benchmark.
<img width="1468" alt="Screenshot 2022-10-02 at 23 10 59" src="https://user-images.githubusercontent.com/266271/193476465-eeb93447-69f8-4aa0-bd8f-9a5a57157125.png">

Reply:
And I decided to try the following trick:

```
        if (RequiresJSONStringEscaping(string.value))
        {
            std::string escaped;
            // just a guess that 16 bytes for escaped characters will be enough to avoid
            // reallocations
            escaped.reserve(string.value.size() + 16);
            EscapeJSONString(string.value, escaped);

            write(escaped);
        }
        else
        {
            write(string.value);
        }
```

And it seems to be even better:
<img width="1083" alt="Screenshot 2022-10-02 at 23 18 17" src="https://user-images.githubusercontent.com/266271/193476772-a5dd94a5-fb5b-4c50-a546-ea2d05dcc428.png">

Reply:
The `String` performance is worse in the final one?
In any case, the deltas seem quite minor and the optimisation code is relatively small, so we can go with whichever is best.

Reply:
Hm, interesting, tbh I was paying attention to Vector only. Will try to re-check that and will make a final decision. 👍🏻

Reply:
After re-running it several times with and without this preallocation it is really hard to say where `String` is better, so I decided to not waste time on it and just leave existing implementation.